Gutenberg Jetpack Preset: Generate imports dynamically from index.json

ockham opened the pull request:
Alternative to #29410

#### Changes proposed in this Pull Request

Generate the information needed for Calypso's `refreshRegistrations()` dynamically from `index.json`. This is step 1 in the de-duplication program: `editor.js` no longer duplicates information found in `index.json`.

#### Testing instructions

- Navigate to `http://calypso.localhost:3000/block-editor/post/<your-site>`. Verify that the correct blocks are still available from the block picker. (Try with both a Free and Premium plan -- the Simple Payments  block should only be available in the latter.) Also verify that blocks still work
- Build the blocks for Jetpack, and verify the same there

#### Follow-up (for subsequent PRs)

Rename the preset's `editor.js` to `index.js`. Create a new `editor.js` to import `index.js`, and register all blocks there. Use that as entry point for the built bundles. This should allow us to drop any additional logic around editor scripts from `bin/sdk/gutenberg.js` and make it almost entirely generic again.

ockham commented on client/gutenberg/extensions/presets/jetpack/editor.js:
```diff
@@ -18,6 +19,14 @@ import * as TiledGalleryBlock from 'gutenberg/extensions/tiled-gallery';
+		import( '../../' + slug ).then( ( { name, settings } ) => ( { name, settings } ) )
+	);
+
+	return await Promise.all( promises ); // Y U NO RETURN ARRAY BUT PROMISE?
```
Try adding `console.log( getExtensions() )` below this function definition -- it'll spit out a promise instead of an array of block definitions (i.e. `{ name, settings }` pairs). This means we can't simply use `getExtensions` inside of `refreshRegistrations()` to iterate over. (I've also tried making that function `async` and to `await` `getExtensions()`, but no dice).

Does anyone have a better clue about promises/async functions?

ockham replied:
> I've also tried making that function `async` and to `await` `getExtensions()`, but no dice

Hmm, on second try, this seems to do the trick. I'll push a commit.

enejb commented on client/gutenberg/extensions/presets/jetpack/utils/refresh-registrations.js:
```diff
@@ -46,7 +49,18 @@ export default function refreshRegistrations() {
 			} else if ( ! available && registered ) {
+				if ( children ) {
+					children.forEach(
+						( { name: childName } ) => void unregisterBlockType( `jetpack/${ childName }` )
```
We need to be careful about this since. Since if the block is being used somewhere else this could introduce bugs. 
I think we don't have to unregister the inner blocks. They will not show up i the selector if the container block in which the reside is not available.

ockham replied:
> We need to be careful about this since. Since if the block is being used somewhere else this could introduce bugs.

Oh, like when a block can have multiple parents? Hmm, good point.

> I think we don't have to unregister the inner blocks. They will not show up i the selector if the container block in which the reside is not available.

But that might cause duplicated registration when re-registering a parent block :slightly_frowning_face: Maybe I can add a check for a blocks parent's upon unregistering, to make sure that it has no other potential parents.

ockham replied:
2563f12

enejb commented on client/gutenberg/extensions/contact-form/editor.js:
```diff
@@ -4,7 +4,7 @@
  */
 import registerJetpackBlock from 'gutenberg/extensions/presets/jetpack/utils/register-jetpack-block';
-import { fields, name, settings } from '.';
+import { children, name, settings } from '.';
```
Can we rename this to InnerBlocks instead?

ockham replied:
I was hoping to maybe re-use the same nomenclature for presets later. But since we're not using them that way yet, fine to rename and re-visit later :+1:

tyxla replied:
Yeah, I think `children` can be a bit confusing. `allowedBlocks` (which mirrors the prop name of `<InnerBlocks />`) or `nestedBlocks` also make sense to me. We can also link to https://github.com/WordPress/gutenberg/tree/master/packages/editor/src/components/inner-blocks for context

enejb replied:
I wouldn't want to use `allowedBlocks` because we allow more then just the fields to be added to the contact form block.

ockham replied:
`nestedBlocks` sounds kinda good to me, tho :thinking:

ockham replied:
Or maybe just `childBlocks`? I mean they do reference their `parent`...

ockham replied:
3b7b2a7 -- I settled on `childBlocks`. `InnerBlocks` seems to refer more to the specific component used for that, `nestedBlocks` seems to indicate more the abstract concepts (possibly including multiple layers of nesting).

tyxla replied:
`childBlocks` is great 👍 nice!

enejb commented:
I was able to test this and it looks like we are now loading a bunch more JS files. 
Each block is being split into its own js file. 

This is the network tab (js files only)
This Branch:
<img width="617" alt="screen shot 2018-12-17 at 4 42 32 am" src="https://user-images.githubusercontent.com/115071/50088054-cf978480-01b6-11e9-96b6-78cb9ea57128.png">

Current master:
<img width="607" alt="screen shot 2018-12-17 at 4 46 55 am" src="https://user-images.githubusercontent.com/115071/50088068-d625fc00-01b6-11e9-8f40-f9281cb62dd2.png">

Other then that this worked as expected.

sirreal commented on client/gutenberg/extensions/presets/jetpack/editor.js:
```diff
@@ -5,29 +5,22 @@
+
+export async function getExtensions() {
+	const promises = extensionSlugs.map( slug =>
+		import( '../../' + slug ).then( ( { children, name, settings } ) => ( {
```
To reduce the extra imports and have webpack resolve and bundle in a single chunk, you should be able to set the mode:

https://webpack.js.org/api/module-methods/#magic-comments
```suggestion
		import(   /* webpackMode: "eager" */ '../../' + slug ).then( ( { children, name, settings } ) => ( {
```

ockham commented:
Thanks @enejb, @tyxla, and @sirreal for your reviews and suggestions! I've addressed all feedback and will merge as soon as CircleCI goes green. (If you're unhappy with any of my choices, let me know and I'll follow up with a fix.)

enejb commented:
Nicely done!

tyxla commented:
Thanks for this, Bernie!
